### ISSUE TITLE
ContentNodeSlim "popular" view can return an empty string as response

### ISSUE THREAD

**Issue by jonboiser** [learningequality/kolibri#5456]
### Observed behavior

Related to #5407, a request to `/api/content/contentnode_slim/popular/?by_role=true&include_fields=num_coach_contents&contentCacheKey=1556915747`

Is returning a 200 response with `""` on my local version, which in turn causes the problem in #5407. The other detail view ("resume" and "next_steps") are working correctly.

**In the Network inspector in chrome dev tools**

<img width="638" alt="Screen Shot 2019-05-06 at 11 57 22 AM" src="https://user-images.githubusercontent.com/10248067/57248244-c3da1c80-6ff6-11e9-85a1-8cec909f6247.png">

I do not know what, specifically, is causing the "popular" view to return nothing.
 

…

### Expected behavior

Assuming, this is a legitimate response to, for example, not having any "popular" resources, a call to this endpoint returns something acceptable to the Api Resource class, like `[]`.

### User-facing consequences

Servers with the same conditions as mine (whatever they are) will not have a working "Learn > Recommended"  page.

### Errors and logs

…

### Steps to reproduce

…

### Context
0.12. develop

**Comment by jonboiser** [comment]
Inspecting this line in the debugger:

https://github.com/learningequality/kolibri/blob/cf19c722c067a102a415577058b1d95c25dab7f5/kolibri/core/content/api.py#L609

In my instance, `cache.get(cache_key)` is equal to None.

I don't see similar code in Next Steps or Recommended views, but see a slightly different cache-related code for the "ancestors" view (with a guard against the cached value being None):

https://github.com/learningequality/kolibri/blob/cf19c722c067a102a415577058b1d95c25dab7f5/kolibri/core/content/api.py#L461-L471

As an experiment, I changed this line in the popular view:

```diff
-       return Response(cache.get(cache_key))
+        return Response(serializer.data)
```

which fixed the problem for me.

**Comment by rtibbles** [comment]
When you returned the serializer.data what was the value, an empty list?

**Comment by jonboiser** [comment]
No, `serializer.data` was an array of contentnodes.

**Comment by jonboiser** [comment]
I'm seeing the same thing on the ancestors view

https://github.com/learningequality/kolibri/blob/cf19c722c067a102a415577058b1d95c25dab7f5/kolibri/core/content/api.py#L457-L472

Typing `cache.get(cache_key) is None` returns True, but in this case the ancestors API returns `Response(serialzier.data)`, so the Topics pages still work fine.

Is there something off about how the cache is set up locally, maybe?

**Comment by rtibbles** [comment]
We are using the Django null cache, which is meant to just accept sets, and then return None for gets. Maybe it's returning empty strings for fun?

**Comment by rtibbles** [comment]
Should just be returning the default (which as we are not passing the kwarg, should be None): https://github.com/django/django/blob/stable/1.11.x/django/core/cache/backends/dummy.py#L18

**Comment by jonboiser** [comment]
So what do lines like `Response(cache.get(cache_key)) ` mean if `cache.get` returns None?

**Comment by rtibbles** [comment]
They shouldn't be being invoked, as the cache is returning `None` and so shouldn't go down that branch, no?

**Comment by jonboiser** [comment]
For "ancestors", it seems like the `if cache.get(cache_key) is not None: ` is dead code, but for the "popular" view `return Response(cache.get(cache_key))` is always returned (but the intention is that the cache will be set immediately before this line if the cached value is None)

**Comment by rtibbles** [comment]
The ancestors view line is not dead code - we only use the dummy cache in development, in production we use a local memory cache.

The issue is probably that the dummy cache is returning None, so we need to not use the 'set cache then read from cache' idiom, because it will break in development mode because it always returns None.

**Comment by jonboiser** [comment]
Or use `Response(cache.get(cache_key, default=serialized_data))`

**Comment by rtibbles** [comment]
Right, either would work.

**Comment by jonboiser** [comment]
Implemented the solution above in #5486. In local development mode, the use-the-cache branch should always be bypassed, and the view returns `Response(serializer.data)`.
